hy/compiler.py

```diff
@@ -598,6 +621,166 @@ def compile_raise_expression(self, expr, root, exc, cause):
             expr, type=ret.expr, exc=exc,
             inst=None, tback=None, cause=cause)
 
+    notsym = lambda *dissallowed: some(
+        lambda x: isinstance(x, Symbol) and str(x) not in dissallowed
+    )
+    keepsym = lambda wanted: some(lambda x: isinstance(x, Symbol) and x == wanted)
+    _match_value = SYM | LITERAL
+    _pattern = forward_decl()
+    _assignment = maybe(sym(":as") + SYM)
+    _pattern.define(
+        _match_value + _assignment
+        | brackets(many(_pattern | unpack("iterable"))) + _assignment
+        | pexpr(keepsym("|"), many(_pattern)) + _assignment
+        | pexpr(keepsym(","), many(_pattern)) + _assignment
+        | curly_brackets(many(LITERAL + _pattern), maybe(pvalue("unpack-mapping", SYM))) + _assignment
+        | pexpr(
+            notsym("|", ",", "unpack-mapping", "unpack-iterable"),
+            many(_pattern),
+            many(KEYWORD + _pattern),
+        ) + _assignment
+    )
+    match_clause = _pattern + maybe(sym(":if") + FORM) + _assignment
+
+    @special(
+        (PY3_10, "match"), [FORM, oneplus(pexpr(sym("case"), match_clause, many(FORM)))]
+    )
+    def compile_match_expression(self, expr, root, target, clauses):
+        subject = self.compile(target)
+        return_var = asty.Name(expr, id=mangle(self.get_anon_var()), ctx=ast.Store())
+
+        has_wildcard_pattern = False
+        lifted_if_defs = []
+        match_cases = []
+        for *pattern, guard, bad_assignment, body in clauses:
+            if pattern[1] and bad_assignment:
+                raise self._syntax_error(bad_assignment, "duplicate :as assignment form in match clause")
+
+            if bad_assignment:
+                raise self._syntax_error(bad_assignment, ":as assignment form cannot come after an :if guard")
+
+            body = self._compile_branch(body)
+            body += asty.Assign(pattern[0], targets=[return_var], value=body.force_expr)
+            body += body.expr_as_stmt()
+            body = body.stmts
+
+            if pattern[0] == Symbol("_"):
```

Comment:
`case catchall: ...` where `catchall` is a valid symbol is allowed in match statements to name the catchall case, so this should be
```python
```

Reply:
fixed